feat: at_chops changes for pkam using private key from secure element

murali-shris opened the pull request:
**- What I did**
 - changes to support pkam using private key from secure element
 - analyser changes
 - bug fix in encoding/decoding of signature
**- How I did it**
- introduced a public method readPublicKey(String publicKeyId) in at_chops_base. Clients using secure element will be extending AtChopsImpl and override readPublicKey(...) method
- provided setters for encryption key pair and symmetric key. If these keys are not available during at_chops creation, they can be set later on.
- added base64Encoding to the recently added method sign(..). Modified documentation for base64encoding/base64decoding
**- How to verify it**
n/a since no changes in implementation in at_chops

murali-shris committed:
feat: at_chops changes for pkam using private key from secure element

gkc commented on packages/at_chops/lib/src/at_chops_base.dart:
```diff
@@ -108,4 +110,8 @@ abstract class AtChops {
   /// Refer to [DefaultHash] for default implementation of hashing.
   String hash(Uint8List signedData, AtHashingAlgorithm hashingAlgorithm);
+
+  /// Reads a public key from SIM card based on ID. SIM card should have an IOTSafe compliant app
```
From SIM is only one example of what an implementation could do.

murali-shris replied:
please check whether my latest docs are fine.

gkc commented on packages/at_chops/lib/src/at_chops_impl.dart:
```diff
@@ -361,4 +361,10 @@ class AtChopsImpl extends AtChops {
+
+  @override
+  String readPublicKey(String publicKeyId) {
+    // This method is implemented only for extensions of AtChops that use secure element for private keys
```
Again this is just one example use (the main example for sure) but I could choose to implement this in some other way also

murali-shris replied:
please check whether my latest docs are fine.

gkc reviewed (changes requested):
Docs could be made more precise

murali-shris committed:
fix: review comments and dart format files

gkc commented on packages/at_chops/lib/src/algorithm/ecc_signing_algo.dart:
```diff
@@ -29,7 +29,8 @@ class EccSigningAlgo implements AtSigningAlgorithm {
     if (publicKey == null && _privateKey == null) {
-      throw AtSigningVerificationException('public key not set not ecc verification');
+      throw AtSigningVerificationException(
+          'public key not set not ecc verification');
```
1. I don't really understand the exception message here. Should it be, "Cannot verify signature - publicKey not provided or privateKey not available"
2. Should it be an `||` condition rather than an `&&`?

murali-shris replied:
In ECDSA, public key can be derived from the private key.
E.g pubKey = _privateKey!.publicKey;
There are two ways verify methods can be called.
1) final eccAlgo = EccSigningAlgo();
eccAlgo.privateKey = //some ecc private key;
privateKey will be used for signing.While verifying signatue, publickey will be derived from the private key and used for verification
2) Zariot use case.Private is not available to us. 
So we will invoke EccSigningAlgo().verify(data, signature, //pass the public key)
I am throwing an exception if both are not available to verify method. if it is || then it will always be true. Hope my explanation is correct?

gkc replied:
Got it - makes sense. Can you update the exception message as I suggested please?

murali-shris replied:
done

gkc commented on packages/at_chops/lib/src/at_chops_impl.dart:
```diff
@@ -364,7 +364,7 @@ class AtChopsImpl extends AtChops {
   @override
   String readPublicKey(String publicKeyId) {
-    // This method is implemented only for extensions of AtChops that use secure element for private keys
+    // This method is implemented only for extensions of AtChops that use secure element or any other source for private keys apart from the default source(.atKeys file)
```
`other than` rather than `apart from`

murali-shris replied:
changed.

murali-shris commented:
Found a bug while testing. We replaced at_chops.signString(..) method with at_chops.sign(..) in at_lookup(yet to be merged). signString was base64 encoding the pkam signature..That is missing in sign method. Will fix this bug in this PR.

murali-shris committed:
fix: base64encoding bug, test case changes and analyser fixes

murali-shris commented:
> Found a bug while testing. We replaced at_chops.signString(..) method with at_chops.sign(..) in at_lookup(yet to be merged). signString was base64 encoding the pkam signature..That is missing in sign method. Will fix this bug in this PR.

Fixed the issue. Ran the changes with at_client and at server
https://github.com/atsign-foundation/at_client_sdk/pull/965
https://github.com/atsign-foundation/at_server/pull/1267 (end2test failure due to test flakiness - not related to at_chops changes)

gkc reviewed (changes requested):
LGTM except for an exception message that needs to be made clearer

murali-shris committed:
fix: change exception message

gkc reviewed (approved):
LGTM
